src/v/utils/chunked_kv_cache.h

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "container/chunked_hash_map.h"
+#include "utils/s3_fifo.h"
+
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/util/optimized_optional.hh>
+
+#include <boost/intrusive/list.hpp>
+#include <boost/intrusive/options.hpp>
+#include <boost/intrusive/parent_from_member.hpp>
+
+namespace utils {
+
+/**
+ * A basic key-value cache implementation built on top of the s3_fifo::cache.
+ */
+template<
+  typename Key,
+  typename Value,
+  typename Hash = std::conditional_t<
+    detail::has_absl_hash<Key>,
+    detail::avalanching_absl_hash<Key>,
+    ankerl::unordered_dense::hash<Key>>,
+  typename EqualTo = std::equal_to<Key>>
+class chunked_kv_cache {
+    struct cached_value;
+    struct evict;
+    using cache_t = s3_fifo::cache<
+      cached_value,
+      &cached_value::hook,
+      evict,
+      s3_fifo::default_cache_cost>;
+
+public:
+    using config = cache_t::config;
+
+    explicit chunked_kv_cache(config config)
+      : _cache{config, evict{*this}} {}
+
+    ~chunked_kv_cache() noexcept = default;
+
+    // These contructors need to be deleted to ensure a stable `this` pointer.
+    chunked_kv_cache(chunked_kv_cache&&) = delete;
+    chunked_kv_cache& operator=(chunked_kv_cache&&) noexcept = delete;
+    chunked_kv_cache(const chunked_kv_cache&) = delete;
+    chunked_kv_cache& operator=(const chunked_kv_cache&) noexcept = delete;
+
+    /**
+     * Inserts a value for a given key into the cache.
+     *
+     * Returns true if the value was inserted and false if there was already a
+     * value for the given key in the cache.
+     */
+    bool try_insert(const Key& key, ss::shared_ptr<Value> val);
+
+    /**
+     * Gets the key's corresponding value from the cache.
+     *
+     * Returns std::nullopt if the key doesn't have a value in the cache.
+     */
+    ss::optimized_optional<ss::shared_ptr<Value>> get_value(const Key& key);
+
+    using cache_stat = struct cache_t::stat;
+    /**
+     * Cache statistics.
+     */
+    struct stat : public cache_stat {
+        /// Current size of the cache index.
+        size_t index_size;
+    };
+
+    /**
+     * Returns the current cache statistics.
+     */
+    [[nodiscard]] stat stat() const noexcept;
+
+private:
+    using ghost_hook_t = boost::intrusive::list_member_hook<
+      boost::intrusive::link_mode<boost::intrusive::safe_link>>;
+
+    struct cached_value {
+        Key key;
+        ss::shared_ptr<Value> value;
+        s3_fifo::cache_hook hook;
+        ghost_hook_t ghost_hook;
+    };
```

Comment:
if you wanted to optimize this to avoid an allocation, you could ask the user to put these hooks into the values that they are storing in the map. it probably goes against the design as a general use key-value store, but its just something to know is on the table later.

Reply:
I had a similar thought. Instead of having a user pass a The tricky part of doing this is that the ghost fifo hooks would still need to exist after the value has been released/freed by the cache. 

It'd be nice if `boost::intrusive::list_member_hook` implemented a move constructor then we could use it in the `chunked_hash_map` without worrying about having it in some stable location. I.e, with a `std::unique_ptr` allocation like I'm doing.

I might go ahead and subclass it actually and implement move constructors for it. It should be relatively simple. Then the allocation can be removed without any additional effort on the users.